### PR TITLE
[MM-59911] Force individual users to receive the `user_added` and `group_added` events instead of relying on the channel membership

### PR DESCRIPTION
**Pull request by devinbinnie**
#### Summary
Over the past few months I was investigating some issues related to cached channel membership in the websocket connections across clusters. This resulted in these two changes:
- https://github.com/mattermost/mattermost/pull/27013 which forced the `user_added` event to be dispatched to the newly added user regardless of channel membership
- https://github.com/mattermost/mattermost/pull/27204 which undid the above after we found that the invalidate cache message just wasn't being sent across the cluster. This changes also forces us to make sure we go to the master DB when getting the channel memberships.

However, it seems people are still missing unreads due to being added to new channels in clustered environments. After some testing I've found that there are still cases where the cache invalidation will not happen if a user's websocket is on a different node than the one that added them as a channel member, before the websocket event is published to that same node. So we have a race condition.

In order to deal with this I've brought back the code from the original change that forces the `user_added` (and in this case the `group_added` event) to be sent directly to the user that was added to the channel (or when the group message is created, send it to each added member). This will ensure that the cache for channel members will not get in the way of users potentially missing websocket events, thus creating missed messages.

I'm not sure if there's a more elegant solution, other than perhaps having a more aware cluster messaging system that knows we need to wait for the cache to be invalidated before we can send the websocket events, but I don't see much application to that outside of this case.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-59911

```release-note
Fixed another issue where users may not see channels they were added to/messages from those channels in clustered environments
```

### CODE REVIEW

**Comment by agnivade** [comment]
> I'm not sure if there's a more elegant solution, other than perhaps having a more aware cluster messaging system that knows we need to wait for the cache to be invalidated before we can send the websocket events, 

In fact, we do have something like that: https://github.com/mattermost/mattermost/blob/1158e6358c2dbc2a3eeeb0a72df78b7209f11466/server/public/model/cluster_message.go#L69

If you set that field to true for the user cache invalidation event, then it will wait until the message has been sent.

But again, there's still the edge-case that it only waits for the request to be sent, and does not know whether actually all the nodes has processed it or not. I don't think there is any way around this except changing the paradigm to a request-response model - which we do have in fact for things like `GetClusterLogs`, `GetPluginInfo` etc. 

It shouldn't be that hard to come up with a sync version of the user cache invalidation message. But I think that's an overkill.

@streamer45 @isacikgoz - Wondering if you have any thoughts on this. It's an interesting edge-case.

**Comment by streamer45** [comment]
> It shouldn't be that hard to come up with a sync version of the user cache invalidation message. But I think that's an overkill.

@agnivade Would that be a non-issue in a Redis-backed deployment?

**Comment by streamer45** [comment]
> Indeed. That would be a non-issue, because there are no cluster invalidation messages in Redis. We invalidate the cache directly, and it's synchronous as you have to wait for the response to come back. Redis just clears away this whole class of race conditions.

Cool, then the workaround proposed here seems reasonable as we'd likely want to push our enterprise customers to consider Redis whenever feasible. And we could highlight an increased reliability as much as performance.

/cc @nab-77

**Comment by agnivade** [comment]
Sounds good. Additionally, I also think setting `WaitForAllToSend` for some of the most high frequency cluster message types would also help in throttling traffic, which as we have seen can become spiky at very high scale.

**Review by agnivade — APPROVED** [review]

File: server/channels/app/channel.go [hunk 1]
```diff
@@ -499,11 +499,11 @@ func (a *App) CreateGroupChannel(c request.CTX, userIDs []string, creatorId stri
 
 	for _, userID := range userIDs {
 		a.InvalidateCacheForUser(userID)
-	}
 
-	message := model.NewWebSocketEvent(model.WebsocketEventGroupAdded, "", channel.Id, "", nil, "")
-	message.Add("teammate_ids", model.ArrayToJSON(userIDs))
-	a.Publish(message)
+		message := model.NewWebSocketEvent(model.WebsocketEventGroupAdded, "", channel.Id, userID, nil, "")
+		message.Add("teammate_ids", model.ArrayToJSON(userIDs))
```
Comment (BenCookie95) [inline server/channels/app/channel.go:504]: 0/5 but you could just do this outside the loop `userIDsJSON := model.ArrayToJSON(userIDs)`

**Review by BenCookie95 — APPROVED** [review]

**Pull request merged**